### PR TITLE
feat(config): Add configurable toolchain profiles

### LINKED ISSUE #673: Configurable toolchain
### Overview

Unikraft is a highly modular library operating system designed for the cloud.  It's high degree of modularization allows for extreme customization and specialization. As such, its tooling should not interfere with the user's desire to support such customization. Towards increasing the unikernel's developer's ability to customize the build whilst simultaneously systemitizing the process of retrieving, organizing and generally facilitating the build of a unikernel based on Unikraft and its many components, the supported tooling, `kraft`, should allow for the injection of the user's environment and or additional toolchain requirements. 

### Feature request summary

This project is designed to better facilitate the dynamic injection of user provided variables into Unikraft's build system through the addition of a dynamically configured `toolchain` towards greater customization of the unikernel build through the use of its command-line companion client tool, `kraft`. This manifests itself as an injection into KraftKit's core configuration system and must propagate across the codebase appropriately.

Distinct results of this addition would enable, but are not limited to: alternating the GNU Compiler Collection (GCC) version, providing alternative compile-time flags, and more.

This project tracks the addition of a map to the internal configuration system entitled `toolchain` which sets the values of programs that are ultimately passed to Unikraft's build system.

The addition is first made as an entry to the [`KraftKit`](https://github.com/unikraft/kraftkit/blob/staging/config/config.go#L17) config structure in 
```diff
  type KraftKit struct {
    [...]
+   Toolchain map[string]string `yaml:"toolchain" long:"toolchain" usage:"Set specific global programs in Unikraft's build system"
    [...]
  }
```

These values need to be passed any `make` invocations as in-line variables via the option [`make.WithVar`](https://github.com/unikraft/kraftkit/blob/staging/make/options.go#L116) in relevant parts of KraftKit's commands that interact with the build, namely in [`cmd/kraft/build/build.go`](https://github.com/unikraft/kraftkit/blob/staging/cmd/kraft/build/build.go) since it is in these files that access to the map is possible via `config.G[config.KraftKit](ctx).Toolchain`.

Additionally, the specific entry for `CC` should be added as default during the installation process, i.e. the user should be asked whether they want to change the default value for `CC` which should be set to `gcc`.

A sample change to the user's configuration could look something like:

```yaml
toolchain:
  CC: /path/to/alternative/gcc
  UK_CFLAGS: -fdiagnostics-color=always
```

~~The same top-level `toolchain` should also be added as a directive/element into the `Kraftfile`, with the ability to specify it per-target.~~

Such configuration changes, when applied to KraftKit's configuration system, would allow for the dynamic injection of toolchain arguments via command-line flags, e.g. during the the construction of a unikernel binary image:

```
kraft --toolchain.UK_CFLAGS="-fdiagnostics-color=always" build
```

### Additional context

Such a configuration system will enable KraftKit users to pass in additional flags to Unikraft's build system following the merge of https://github.com/unikraft/unikraft/pull/957

```[tasklist]
### Steps
- [x] Test building an application via `make`
- [x] Try passing different toolchain options to the `make` command
- [ ] Try to hardcode-inject a toolchain option within the kraftkit `make` package. See if it works as expected, or needs redesigning
- [ ] Introduce the Toolchain global config variable as hinted above
- [ ] ~~Introduce the toolchain kraftfile element (see https://github.com/unikraft/kraftkit/pull/1463 https://github.com/unikraft/kraftkit/pull/1286 for suggestions how to do things)~~
- [ ] Introduce cli options/commands to set/list toolchain variables
```

### PR DESCRIPTION
**Pull request by robaa12**
### Prerequisite checklist
  
  - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes
to the project;
  - [x] Tested your changes locally.
  
  ### Description of changes
  
  This PR implements the initial configuration and build-merging infrastructure to support configurable toolchain
profiles (Closes #673). It introduces `ToolchainProfiles` to allow users to define named compiler/build environments
in their global configuration, and implements a 3-level toolchain merging precedence.
  
  The toolchain precedence resolves as:
  
  
CLI flags (e.g., -T CC=clang) > Selected Profile (e.g., --toolchain-profile=clang-debug) > Global Config Defaults
(e.g., toolchain.CC)

  
  
  | Component / File | Description of Changes |
  |---|---|
  | **`config/config.go`** | Added `ToolchainProfiles map[string]map[string]string` to the global configuration schema.
  | **`config/manager.go`** | Refactored `Set`/`Unset` methods to use a unified `traverse` helper, and added recursive validation/mutation for two-level nested maps. |
  | **`internal/cli/kraft/system/list/list.go`** | Updated `flattenConfig` to recursively format and print nested map elements (making them visible via `kraft system list`). |
  | **`internal/cli/kraft/build/build.go`** | Added the `--toolchain-profile` CLI flag to `BuildOptions` and updated `mergeToolchain` to enforce the 3-level precedence hierarchy. |
  | **`internal/cli/kraft/build/builder_kraftfile_unikraft.go`** | Propagated the selected profile options to the underlying GNU Make options generator. |
  | **`config/manager_test.go`** | Added unit test coverage for setting, unsetting, and edge cases (e.g., too-deep paths) of nested maps. |
  | **`internal/cli/kraft/build/build_toolchain_test.go`** | Updated toolchain unit tests to verify precedence resolution across global defaults, profiles, and CLI overrides. |
  
  ### Verification
  
  All unit tests compiled and passed locally:
  - `go test ./config/...`
  - `go test ./internal/cli/kraft/build/...`

### CODE REVIEW

**Comment by robaa12** [comment]
@craciunoiuc  If you have some free time, could you review it?

File: config/manager.go [hunk 1]
```diff
@@ -256,60 +239,116 @@ func (cm *ConfigManager[C]) Set(key string, val any) error {
...
 		}
+		return nil
 	}
 
+	if !field.CanSet() {
+		return errors.New("cannot unset field: " + keys[i])
+	}
+	field.Set(reflect.Zero(field.Type()))
 	return nil
 }
```
Comment (ijaidev) [inline config/manager.go:299]: Could we keep Set and Unset at their places and move the new traverse helper below them, so the visible reordering/splitting of Set/Unset is avoided? 
That will make the change smaller and easier to review/merge.

File: internal/cli/kraft/build/build.go [hunk 2]
```diff
@@ -412,11 +414,18 @@ func moveFile(src, dst string) error {
...
 	result := map[string]string{}
 	for k, v := range global {
 		result[k] = v
 	}
+	if profileName != "" {
+		if profile, ok := config.G[config.KraftKit](ctx).ToolchainProfiles[profileName]; ok {
+			for k, v := range profile {
+				result[k] = v
+			}
+		}
```
Comment (ijaidev) [inline internal/cli/kraft/build/build.go:431]: I would throw an error to the user if the given profile doesn't exist.

**Review by ijaidev — CHANGES_REQUESTED** [review]
Could you please merge/squash commits `4e101d7` and `34222bb`? Also, I'd prefer not to change the function definitions. Please keep the signatures as they are, update only the internal logic, and extract the common logic into a helper function at the bottom to keep the diff clean.

Otherwise, LGTM. 🙏

**Comment by ijaidev** [comment]
You can verify it after squashing those commits

File: config/manager.go [hunk 3]
```diff
@@ -189,82 +189,121 @@ func getYAMLTag(field reflect.StructField, ok bool) string {
...
 
-	v = v.Elem() // Dereference the pointer
-	keys := strings.Split(key, ".")
+	// Handle map[string]map[string]string fields.
+	if field.Kind() == reflect.Map && field.Type().Elem().Kind() == reflect.Map {
+		if i+2 != len(keys)-1 {
+			return errors.New("invalid nested map key path length: " + keys[i])
+		}
+		profileKey := keys[i+1]
+		varKey := keys[i+2]
```
Comment (ijaidev) [inline config/manager.go]: I would use generic names like `innerKey`, `outerKey` to avoid confusion.

File: config/manager.go [hunk 4]
```diff
@@ -189,82 +189,121 @@ func getYAMLTag(field reflect.StructField, ok bool) string {
...
+		return err
+	}
+
+	// Handle map[string]map[string]string fields.
+	if field.Kind() == reflect.Map && field.Type().Elem().Kind() == reflect.Map {
+		if i+2 != len(keys)-1 {
+			return errors.New("invalid nested map key path length: " + keys[i])
+		}
+		profileKey := keys[i+1]
+		varKey := keys[i+2]
```
Comment (ijaidev) [inline config/manager.go]: >I would use generic names like `innerKey`, `outerKey` to avoid confusion.

-----
Same here

**Review by ijaidev — CHANGES_REQUESTED** [review]
A few more changes to make it more robust and we are good to go.

**Review by ijaidev — APPROVED** [review]
LGTM!
```
Reviewed-by: Jaidev Singh <hi@jaidev.me>
Approved-by: Jaidev Singh <hi@jaidev.me>
```

**Review by craciunoiuc — APPROVED** [review]
All good here. Thanks!

```
Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.com>
Approved-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.com>
```
